File: trunk/api-java/src/org/glite/rgma/OnDemandProducer.java
/*
 * Copyright (c) Members of the EGEE Collaboration. 2004.
 * See http://eu-egee.org/partners/ for details on the copyright holders.
 * For license conditions see the license file or http://eu-egee.org/license.html
 */

package org.glite.rgma;

import java.util.ArrayList;
import java.util.List;

/**
 * A client uses an on-demand producer to publish data into R-GMA when the cost of creating each message is high. The
 * on-demand producer only generates messages when there is a specific query from a consumer.
 */
public class OnDemandProducer extends Producer {

	private String m_hostName;
	private int m_port;
	private List<Table> m_tables = new ArrayList<Table>();

	/**
	 * Creates an on-demand producer.
	 * 
	 * @param hostName
	 *            the host name of the system that will respond to queries
	 * @param port
	 *            the port on the specified host that will respond to queries
	 * @throws RGMAPermanentException
	 * @throws RGMATemporaryException
	 */
	public OnDemandProducer(String hostName, int port) throws RGMAPermanentException, RGMATemporaryException {
		m_servletName = "OnDemandProducerServlet";
		ServletConnection connection = new ServletConnection(m_servletName);
		connection.addParameter("hostName", hostName);
		connection.addParameter("port", port);
		Tuple tuple = connection.sendNonResourceCommand("createOnDemandProducer").getData().get(0);
		m_resourceId = tuple.getInt(0);
		m_hostName = hostName;
		m_port = port;
	}

	/**
	 * Declares a table into which this producer can publish. A subset of a table can be declared using a predicate.
	 * 
	 * @param name
	 *            the name of the table to publish into
	 * @param predicate
	 *            an SQL WHERE clause defining the subset of a table that this producer will publish. To publish to the
	 *            whole table, an empty predicate can be used.
	 * @throws RGMAPermanentException
	 * @throws RGMATemporaryException
	 */
	public void declareTable(String name, String predicate) throws RGMAPermanentException, RGMATemporaryException {
		try {
			doDeclareTable(name, predicate);
		} catch (UnknownResourceException e) {
			try {
				restore();
				doDeclareTable(name, predicate);
			} catch (UnknownResourceException e1) {
				throw new RGMATemporaryException(e1.getMessage());
			}
		}
		m_tables.add(new Table(name, predicate));
	}

	private void doDeclareTable(String name, String predicate) throws RGMAPermanentException, RGMATemporaryException, UnknownResourceException {
		ServletConnection connection = getNewConnection();
		connection.addParameter("tableName", name);
		connection.addParameter("predicate", predicate);
		checkOK(connection.sendCommand("declareTable"));
	}

	private void restore() throws RGMAPermanentException, RGMATemporaryException, UnknownResourceException {
		OnDemandProducer p;
		p = new OnDemandProducer(m_hostName, m_port);
		for (Table t : m_tables) {
			p.doDeclareTable(t.m_name, t.m_predicate);
		}
		m_resourceId = p.m_resourceId;
	}

	private class Table {

		private String m_name;
		private String m_predicate;

		private Table(String name, String predicate) {
			m_name = name;
			m_predicate = predicate;
		}

	}

}

File: api-cpp/src/OnDemandProducer.cpp
#include  "rgma/OnDemandProducer.h"

#include "rgma/ServletConnection.h"
#include "rgma/XMLConverter.h"

namespace glite {
namespace rgma {

OnDemandProducer::~OnDemandProducer() {
}

OnDemandProducer::OnDemandProducer(const std::string & hostName, int port) throw(RGMATemporaryException,
        RGMAPermanentException) :
    Producer("OnDemandProducer"), m_hostName(hostName), m_port(port) {
    m_connection.addParameter("hostName", hostName);
    m_connection.addParameter("port", port);
    TupleSet result;
    m_connection.connect("createOnDemandProducer", result);
    TupleSet::const_iterator it = result.begin();
    m_endPoint.setResourceId(it->getInt(0));
}

void OnDemandProducer::declareTable(const std::string & name, const std::string & predicate)
        throw(RGMATemporaryException, RGMAPermanentException) {
    try {
        doDeclareTable(name, predicate);
    } catch (UnknownResourceException e) {
        try {
            restore();
            doDeclareTable(name, predicate);
        } catch (UnknownResourceException e1) {
            throw RGMATemporaryException(e1.getMessage());
        }
    }
    m_tables.push_back(Table(name, predicate));
}

void OnDemandProducer::doDeclareTable(const std::string & tableName, const std::string & predicate)
        throw(RGMATemporaryException, RGMAPermanentException, UnknownResourceException) {
    clearServletConnection();
    m_connection.addParameter("tableName", tableName);
    m_connection.addParameter("predicate", predicate);
    TupleSet result;
    m_connection.connect("declareTable", result);
    checkOK(result);
}

void OnDemandProducer::restore() throw (RGMAPermanentException, RGMATemporaryException, UnknownResourceException) {
    OnDemandProducer p(m_hostName, m_port);
    tables_iterator ti = m_tables.begin();
    while (ti != m_tables.end()) {
        p.doDeclareTable(ti->m_name, ti->m_predicate);
        ti++;
    }
    m_endPoint.setResourceId(p.m_endPoint.getResourceId());
}

OnDemandProducer::Table::Table(const std::string & name, const std::string & predicate) :
    m_name(name), m_predicate(predicate) {
}

}
}
